src/05-run.md

```diff
@@ -0,0 +1,79 @@
+- `socketwrappermap`: A map of luasocket<->cosock socket pairs
+  - This map is keyed with the table pointer for the luasocket for easily getting back to a cosock socket when you
+    only have a luasocket
+  - This gets populated when a cosock socket is included in a select args table
+- `threaderrorhandler`: Potential error handler function. Not currently settable.
+- `timers`: A list of `timer` tables
+  - A timer has the shape `{timeoutat = "deadline timestamp", callback = "function to be called at deadline", ref = "table pointer"}`
+  - `timers.set`
+    - Updates `timers` to include that value. Also updates a private scoped table named `refs`
+      - `refs` is a map of table pointer<->timer
+  - `timers.cancel`
+    - If the provided table pointer is in `refs`, remove the `callback` and `ref` properties from that table
+    - Set the table pointer key in `refs` to `nil`
+  - `timers.run`
+    - Sort all timeouts by deadline (earliest first)
+    - Pop the timer off the front of the `timers` list
+    - If that `timer.timeoutat` is `nil` or `< socket.gettime()`
+      - Call `timer.callback`
+      - remove this `timer` from `refs`
+    - If there are any more timeouts left, return how long before that timeout should expire
+    - If there are no more timeouts, return `nil`
+  
+## Run loop steps
+
+1. define `wakethreads`
+2. define an empty list of senders (`sendt`), receivers (`recvt`) and a `timeout`
+3. pop all `readythreads` entries into the `wakethreads`
+4. Loop over all threads in `wakethreads`
+   1. If `coroutine.status` for that thread returns "suspended"
+      1. Clear any timers
+      2. Clear any wakers registered with a `timeout`
+      3. `coroutine.resume` with the stored `recv4`, `sendr` and `err` arguments
+      4. if `coroutine.resume` returned `true` in the first position and `coroutine.status` returns "suspended"
+         1. re-populate `threadswaitingfor[thread]` with the 3 other return values from `coroutine.resume`
+            1. These should be the `recvt`, `sendt` and `timeout` values that will populate select args
+         2. set the waker for all sockets in `recvt` and `sendt` to call `wake_thread` and then unset themselves
+         3. if `coroutine.resume` returned a `timeout`, create a new timer for this thread which will call `wake_thread_err` on expirations with the value "timeout"
+      5. if `coroutine.status` returned "dead"
+         1. If `coroutine.resume` returned `false` in the first position and no `threaderrorhandler` has been set
+            1. Raise and error
+               1. If the `debug` library is available, include a `debug.traceback` and the second return value from `cosock.resume`
+               2. else just raise an error with the second return value from `cosock.resume`
+            2. Exit the application
+               1. This calls `os.exit(-1)`
+5. Else, print a warning message if printing is turned on
```

Comment:
Should this be one indentation level deeper? As is it looks like this else is tied to the loop.

Reply:
Yep